api.go

```diff
@@ -95,47 +100,71 @@ func writeDatabaseError(w http.ResponseWriter) {
 	writeError(w, http.StatusInternalServerError, API_ERR_INTERNAL, "database error")
 }
 
-func distributeEvents() {
-	subscribers := make(map[chan string]struct{})
-	mut := sync.Mutex{}
-	msg := ""
-	debouncer := func() {
-		select {
-		case <-time.After(debounceDuration):
-			mut.Lock()
-			subs := subscribers
-			subscribers = make(map[chan string]struct{})
-			m := msg
-			msg = ""
-			incrementETag()
-			mut.Unlock()
-			log.Debugf("Delivering deployment change %s to %d subscribers", m, len(subs))
-			for subscriber := range subs {
-				select {
-				case subscriber <- m:
-					log.Debugf("Handling deploy response for: %s", m)
-					log.Debugf("delivering TO: %v", subscriber)
-				default:
-					log.Debugf("listener too far behind, message dropped")
+func debounce(in chan interface{}, out chan []interface{}, window time.Duration) {
+	send := func(toSend []interface{}) {
+		if toSend != nil {
+			log.Debugf("debouncer sending: %v", toSend)
+			out <- toSend
+		}
+	}
+	for {
+		incoming, ok := <-in
+		if !ok {
+			log.Debugf("closing debouncer")
+			close(out)
+			return
+		}
+		log.Debugf("debouncing %v", incoming)
+		toSend := []interface{}{incoming}
+		for {
```

Comment:
It seems no need for the outer `for`, cause this loop will only execute once.
Once go into inner `for` loop, the program will either stay in the inner loop forever, or return due to channel close.
Try the following code for this `debounce()` function:

	var toSend []interface{} //initialized to nil
	for {
		select {
		case incoming, ok := <-in:
			if ok {
				log.Debugf("debouncing %v", incoming)
				toSend = append(toSend, incoming)
			} else {
				send(toSend)
				log.Debugf("closing debouncer")
				close(out)
				return
			}
		case <-time.After(window):
			send(toSend)
			toSend = nil
		}
	}